[PrivacyPrivilegeManager] Add new API for checking & requesting multiple privileges

xerrni opened the pull request:
### Description of Change ###

C# bindings for checking & requesting multiple privileges using single API call

### API Changes ###
 - TCSACR-184
(It's related Native ACR-1253.)

### Added: ###
#### Class PrivacyPrivilegeManager: ####
Method:
```csharp
public static IEnumerable<CheckResult> CheckPermissions(IEnumerable<string> privileges);
public static Task<RequestMultipleResponseEventArgs> RequestPermissions(IEnumerable<string> privileges);
```
#### Class PermissionRequestResponse: ####
Event:
```csharp
public RequestResult Result { get; internal set; }
public string Privilege { get; internal set; }
```
#### Class RequestResponseEventArgs : EventArgs ####
Property:
```csharp
public PermissionRequestResponse Response { get; }
```
#### Class RequestMultipleResponseEventArgs : EventArgs ####
Property:
```csharp
public CallCause Cause { get; internal set; }
public IEnumerable<PermissionRequestResponse> Responses { get; internal set; }
```

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -120,6 +211,83 @@ public static void RequestPermission(string privilege)
+        /// </code>
+        /// </example>
+        /// <since_tizen> 5 </since_tizen>
+        public static void RequestPermissions(string[] privileges)
```
Why don't you provide async method?

public static Task<RequestResponse> RequestPermissions(string[] privileges)

myroot replied:
And do not use array in API
Use `IEnumerable<string>` instead of `string[]`

myroot replied:
2. Why do you think so?
Please run below code
``` c#
            if (typeof(string[]).IsAssignableFrom(typeof(IEnumerable<string>))) {
                System.Console.WriteLine("true");
            }            
            string[] result = new string[10];
            IEnumerable<string> mylist = result;
```

myroot replied:
Please see .net framework design guide

X DO NOT use ArrayList or List<T> in public APIs.
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections

myroot replied:
And 
`public static void RequestPermission(string privilege)`
It should be deprecated because it is not good usability

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -180,6 +348,91 @@ public static WeakReference<ResponseContext> GetResponseContext(string privilege
+        /// </code>
+        /// </example>
+        /// <since_tizen> 5 </since_tizen>
+        public static WeakReference<MultipleResponseContext> GetResponseContext(string[] privileges)
```
Why do you use WeakReference?

xerrni replied:
1. To prevent memory leaks when handlers that are attached to event sources will not be destroyed at the same time.

Reference:
https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/aa970850(v=vs.100)
2. To be consistent with
```csharp
public static WeakReference<ResponseContext> GetResponseContext(string privilege)
```

myroot replied:
Why do you think so?
It is not meaory leaks, because it released by GC,
and why do you thinks it should be destroyed at the same time?

WeakReference should not used in API

myroot replied:
I know your intent but implemention was wrong

It should be like below changes
https://github.com/myroot/TizenFX/commit/9579961b3931dd52d92cb6e03f8f2a7f3b5a8cb1

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/RequestResponseMultipleEventArgs.cs:
```diff
@@ -0,0 +1,47 @@
+        /// </summary>
+        /// <since_tizen> 5 </since_tizen>
+        public IEnumerable<string> Privileges { get; internal set; }
+    }
```
It is uncomfortable to use. 
Let's see below use case
``` c#
void Handler(RequestMultipleResponseEventArgs arg) {
   var privileges = arg.Privileges.ToList();
   int index = 0;
   foreach (var result in arg.Results) {
       Console.WriteLine($"Result {privileges[index]} : {result}");
       index++;
   }
}
```
Results and Privileges should be a pair becuase it is a related data. 

Consider change to IList from IEnumerable
or Make a some class to combine `Result` and `Privilege`
``` c#
class Response {
    public RequestResult result { get; }
    public string privilege { get; }
}

public class RequestMultipleResponseEventArgs : EventArgs
{
     public CallCause cause { get; }
     public IEnumerable<Response> Responses {get; }
}
public class RequestResponseEventArgs : EventArgs
{
     public CallCause cause { get; }
     public Response Response {get; }
}

```

xerrni replied:
I can change is for `RequestMultipleResponseEventArgs` class.
`RequestResponseEventArgs` was introduced in tizen 4. We don`t want to break backwards compatibility so I could extend it.
```csharp
public class RequestResponseEventArgs : EventArgs
    {
        public CallCause cause { get; internal set; }

        public RequestResult result { get; internal set; }

        public string privilege { get; internal set; }
        public Response Response {get; }
    }
```

WonyoungChoi commented:
Run Build Checker

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -120,6 +237,100 @@ public static void RequestPermission(string privilege)
+        /// </code>
+        /// </example>
+        /// <since_tizen> 5 </since_tizen>
+        public static void RequestPermissions(IEnumerable<string> privileges)
```
developer may not be interested in responseContext.
Subscribing to response is and should be optional.

myroot replied:
Why do you think so? Subscribing is mandatory, if not, developer got a security exception

For example, to access media storage, must grant "http://tizen.org/privilege/mediastorage" privilege
So, developer write a code like below
``` c#
  TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
  var response = PrivacyPrivilegeManager.GetResponseContext("http://tizen.org/privilege/mediastorage");
  PrivacyPrivilegeManager.ResponseContext target;
  response.TryGetTarget(out target);
  target.ResponseFetched += (s, e) =>
  {
      tcs.SetResult(e.result == RequestResult.AllowForever);
   };
   PrivacyPrivilegeManager.RequestPermission("http://tizen.org/privilege/mediastorage");
   bool result = await tcs.Task;
   if (result) {
       // access to media storage
    } else {
        //show error message
   }
```

If you provide as a Task<RequestResult>, developer can make more simple code
``` c#
 public static Task<RequestResult> RequestPermission(string privilege)
```
``` c#
    var result = await PrivacyPrivilegeManager.RequestPermission("http://tizen.org/privilege/mediastorage");
    if (result == RequestResult.AllowForever) {
           // access to media storage
    } else {
           // show error message
    }
```
What do you think about the difference in usability between these two codes?

myroot replied:
And I hope to you distinguish `event` vs `asynchronous result`.
`ResponseFetched` is not event, 
It is a respone of `RequestPermission` so, it is a asynchronus result

`event` can raise more than one. 
`asynchronous result` does not raise if not do request.

`event` has long term lifecycle. it is same with manager object, 
`asynchronous result` has short term lifecycle, it's lifetime was end when request was completed.

And, I think, ResponseFetched should not be implement as event

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -180,6 +391,106 @@ public static WeakReference<ResponseContext> GetResponseContext(string privilege
+            {
+                throw new ArgumentException("Unable to get response context for null argument or 0 privileges.");
+            }
+            string joinedPrivileges = string.Join("", privilegesArray);
```
I need to check privileges count see 473 line: privilegesArray.Length == 0

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -120,6 +237,100 @@ public static void RequestPermission(string privilege)
+            }
+
+            int iterator = 0;
+            for (iterator = 0; iterator < privilegesArray.Length; ++iterator)
```
I still have to check if each Add operation will be successful (307 line).

xerrni replied:
I moved checking privileges argument to separate function as it`s used in 3 places.

myroot replied:
Why not use Overlaps method?
If Overlaps method return true, Add operation never be failed.

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -180,6 +391,106 @@ public static WeakReference<ResponseContext> GetResponseContext(string privilege
+        /// <since_tizen> 5 </since_tizen>
+        public static MultipleResponseContext GetResponseContext(IEnumerable<string> privileges)
+        {
+            string[] privilegesArray;
```
done

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -180,6 +391,106 @@ public static WeakReference<ResponseContext> GetResponseContext(string privilege
+        /// </code>
+        /// </example>
+        /// <since_tizen> 5 </since_tizen>
+        public static MultipleResponseContext GetResponseContext(IEnumerable<string> privileges)
```
And it should not rise when developer set context to null.

myroot replied:
What about this scenario?
``` c#
GetResponseContext(...).ResponseFetched += ....;
GC.Collect();
GC.WaitForPendingFinalizers();
RequestPermissions(...);
//ResponseFetched never raise
```

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -219,7 +531,50 @@ internal ResponseContext(string privilege)
+        /// This class enables having event handlers for multiple privileges.
+        /// </summary>
+        /// <since_tizen> 5 </since_tizen>
+        public class MultipleResponseContext
```
`MultipleResponseContext EventHandler` has to be accessed from outside.

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/Response.cs:
```diff
@@ -0,0 +1,36 @@
+    /// <summary>
+    /// Holds information about privilege request result.
+    /// </summary>
+    public class Response
```
`Response` class name is too common

How about `PermissionRequestResponse`?

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -120,6 +203,101 @@ public static void RequestPermission(string privilege)
+            TaskCompletionSource<RequestMultipleResponseEventArgs> permissionResponsesTask = new TaskCompletionSource<RequestMultipleResponseEventArgs>();
+            int ret = (int)Interop.PrivacyPrivilegeManager.RequestPermissions(privilegesArray, (uint)privilegesArray.Length,
+                        (Interop.PrivacyPrivilegeManager.CallCause cause, Interop.PrivacyPrivilegeManager.RequestResult[] results,
+                        string[] _privileges, uint privilegesCount, IntPtr userData) =>
```
because `privileges` is an argument to upper method.

myroot replied:
Do not use `_` prefix. because it is for field variable.

How about use `string[] requestedPrivileges`

xerrni replied:
done

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -120,6 +203,101 @@ public static void RequestPermission(string privilege)
+            if (ret != (int)Interop.PrivacyPrivilegeManager.ErrorCode.None)
+            {
+                Log.Error(LogTag, "Failed to request permissions.");
+                throw PrivacyPrivilegeManagerErrorFactory.GetException(ret);
```
done

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -112,6 +189,12 @@ public static CheckResult CheckPermission(string privilege)
+                Log.Error(LogTag, "Request for this privilege: " + privilege + " is already in progress.");
+                throw new ArgumentException("Request for this privilege: " + privilege + " is already in progress.");
+            }
+
```
I added it in 290 line

xerrni replied:
I have to throw Exception because delegate `RequestResponseCallback` will not be called.

When `ppm_request_permissions` returns error it just end execution and `RequestMultipleResponseCallback` delegate will not be called.

xerrni commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -120,6 +203,101 @@ public static void RequestPermission(string privilege)
+                        {
+                            Log.Info(LogTag, "Sending request for permissions: ");
+                            RequestMultipleResponseEventArgs requestResponse = new RequestMultipleResponseEventArgs();
+                            PermissionRequestResponse[] permissionResponses = new PermissionRequestResponse[privilegesCount];
```
I used `Array[]` instead of `List` because I know the responses count and they doesn`t change during execution.

myroot commented:
Please fix this issue
https://github.com/Samsung/TizenFX/issues/466

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -203,23 +386,27 @@ internal ResponseContext(string privilege)
+            }
+            ~ResponseContext()
+            {
+                s_responseMap.Remove(_privilege);
```
It has race condition issue, destructor was executed on GC Thread.

and it seem to be dead code, 
because, this object was strongly referenced by s_responseMaps

If _privilege was not existed in s_responseMaps, you do not need to remove from s_responseMap
If _privilege was existed in s_responseMaps, This object never destroyed

xerrni replied:
done

myroot replied:
??

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -203,23 +386,27 @@ internal ResponseContext(string privilege)
                     _ResponseFetched += value;
+                    if (!s_responseMap.ContainsKey(_privilege))
+                    {
+                        s_responseMap[_privilege] = new ResponseContext(_privilege);
```
Why do you create a new ResponseContext?
The result of request will send to new created ResponseContext instead this object, 
You intend it?

xerrni replied:
corrected to 
```csharp
this
```

myroot commented:
Here is **Updated** solution for you
https://github.com/myroot/TizenFX/commit/f26192773abb76aa190a841cad33b46fb94a890c

xerrni commented:
> Here is **Updated** solution for you
> [myroot@f261927](https://github.com/myroot/TizenFX/commit/f26192773abb76aa190a841cad33b46fb94a890c)

Please read my comment under your solution.

myroot commented:
Your solution fail below case
and My solution does  not change API
Did you check my solution?

``` c#
static ResponseContext context = null; PrivacyPrivilegeManager.GetResponseContext("some privilege").TryGetTarget(out context); context.ResponseFetched += some_handler; context.ResponseFetched -= some_handler; // context was removed from s_responseMap PrivacyPrivilegeManager.GetResponseContext("some privilege"); // a new context was created and added in s_responseMap context.ResponseFetched += some_handler; // It will not fired when Response was fetched
```

xerrni commented:
> Your solution fail below case
> and My solution does not change API
> Did you check my solution?
> 
> ```cs
> static ResponseContext context = null; 
> static ResponseContext context2 = null; 
> PrivacyPrivilegeManager.GetResponseContext("some privilege").TryGetTarget(out context); context.ResponseFetched += some_handler; context.ResponseFetched -= some_handler; // context was removed from s_responseMap
> PrivacyPrivilegeManager.GetResponseContext("some privilege").TryGetTarget(out context2);  // a new context was created and added in s_responseMap 
> context.ResponseFetched += some_handler; // It will not fired when Response was fetched
> context2.ResponseFetched += .... ;
> ```

I fixed this.
Can you please no block this Pull Request. It has to be merged before Tizen 5.0 is released.
I will be happy to make another commit with other fixes not related to this Pull request.

myroot commented:
It was not fixed. If you want to merge to master, 
PLEASE, SUBMIT CORRECT CODE.

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -27,22 +29,51 @@ namespace Tizen.Security
+                                }
+                                else
+                                {
+                                    s_responseMap.Remove(privilege);
```
You does not need call Remove, 
becuase, TryGetValue was truning false, it means privilege was not contained into dictionary

myroot commented:
If you want separate a PR from bug fix PR.
DO NOT CHANGE CODE that is not related to Added API

myroot commented on src/Tizen.Security.PrivacyPrivilegeManager/Tizen.Security/PrivacyPrivilegeManager.cs:
```diff
@@ -27,22 +29,51 @@ namespace Tizen.Security
+                        {
+                            try
+                            {
+                                if (s_responseMap.TryGetValue(privilege, out ResponseContext ctx) && ctx != null)
```
`ctx != null` is dead code, 
because, if does not existed ResponseContext, TryGetValue returing false, so `ctx != null` was not executed.
and if existed ResponseContext, ctx can not be null unless you insert null.

myroot commented:
https://github.com/Samsung/TizenFX/issues/466 issue was fixed, 
but code is not good

myroot commented:
Why don't you apply this(https://github.com/myroot/TizenFX/commit/f26192773abb76aa190a841cad33b46fb94a890c) solution? Do you have any problem to apply?

myroot commented:
Please reopen when ready to update
